### DISCUSSION TITLE
Multiple Dear ImGui contexts

### DISCUSSION THREAD

**Discussion by fffffoooooxxxxx** [ocornut/imgui#7578] [New Users Build/Link/Run issues ONLY!]
First, I'm sorry, my english is poor...

Looking directly at the code.

This is my code in render loop:

`   ImGui_ImplDX11_NewFrame();
    ImGui_ImplWin32_NewFrame();
    ImGui::NewFrame();
    
    ImGui::ShowDemoWindow();

    ImGui::Render();
    ImGui_ImplDX11_RenderDrawData(ImGui::GetDrawData());

    ImGui_ImplDX11_NewFrame();
    ImGui_ImplWin32_NewFrame();
    ImGui::NewFrame();

    if (ImGui::Button("button1", ImVec2(50, 50)))
    {

    }

    ImGui::Render();
    ImGui_ImplDX11_RenderDrawData(ImGui::GetDrawData());`

In this code, the click event can't work. How to work good in multiple contexts?

I know "It is STRONGLY preferred that you use docking branch with multi-viewports (== single Dear ImGui context + multiple windows) instead of multiple Dear ImGui contexts.".
But I want to draw multiple windows simultaneously, each with its own click event, and then each window will serve as a 2D texture to render at a certain position in the 3D scene. It's means one imgui-frame will save as a 2D texture, then it will used to render in 3D scene.If I need multiple 2D textures, maybe need to use multiple Dear ImGui contexts, i think so...

**Comment by NostraMagister** [comment]
I use multiple contexts and I added the following at the top of imconfig.h (according ImGui docs):

struct ImGuiContext;
extern thread_local ImGuiContext* ImGuiTLS;
#define GImGui ImGuiTLS

In my code I call:

ImGuiIO& MainWindow::ImGuiInitialize()
{
    // Setup Dear ImGui context
    IMGUI_CHECKVERSION();
    this->gui_context = ImGui::CreateContext();
    
    // Too make ImGui thread safe for contexts we must copy the context to a thread local variable
    ImGuiTLS = this->gui_context;
    
    return ImGui::GetIO();
}

This works in multi-threaded and multiple .dlls in ImGui 1.89.6 and 1.90.6 (i didn't use any other versions).
Works with backends SDL3 3.0.0 and 3.1.2 and Vulkan 1.3.250 and 1.3.283.

Each .dll does need to include the ImGui source and you will see one or more main SDL windows appear in which you can build your ImGui sub-windows. Each .dll will have its OWN SDL main Windows with its OWN event loop. 

Keep the ImGui_ImplSDL3_ProcessEvent in a variable and call it by address otherwise all your instances will be using the same context.

In typedef:  typedef bool (*gui_ProcessEventPtr)(const SDL_Event* event);
In class: gui_ProcessEventPtr       sdl_fProcessEvent = nullptr; 
In code: this->sdl_fProcessEvent = &ImGui_ImplSDL3_ProcessEvent; 

In OLD SDL 3.0.0 you need to do some tweaking if you use filters. I still have that code in SDL 3.1.2 and didn't remove it, but possibly it is no longer needed. SDL 3 gave all events of ALL windows to ALL other windows. So the filter needs to use the window id to discriminate. There is a good chance that this is fixed in SDL 3.1.2., i didn't test it.

In class: SDL_EventFilter           sdl_fEventFilter = nullptr; 
In code: this->sdl_fEventFilter = SDLWindowIdFilter; 

I am in an event-driven system hence I need a filter to let SDL maintain the state and then send events to other queues that consume.
them. One of these queues gives some of the events to ImGui to let it handle the IO related to its windows.

I hope this can help or give you the beginning of a solution for your own code model.
